### PR TITLE
[Fleet] Enable log introspections

### LINKED ISSUE #14452: [Agent]Allow runnings beats to be monitored by Metricbeat.
*Original comment by @ph:*

Starts a Metricbeat sidecar to monitor runnings processes

### PR DESCRIPTION
**Pull request by michalpristas**
Enables log introspection by introducing monitoring component into application
Flow
- `FLEET_MONITORING` program is injected into `AST` with
  - monitoring configuration from agent spec
  - list of programs included in configuration
  - `FLEET_MONITORING` is appended as a last step of configuration change
- operator detect `FLEET_MONITORING` and start/stop respective beats to monitor monitorable beats 
- beats are made monitorable by
  - including logging configuration 
  - including `http.enabled` with `host` specified targeting unix socket

Fixes: #14437 
Fixes #14452

### CODE REVIEW

**Comment by ph** [comment]
#14558 was merged in master and I've rebased the fleet branch with the commit.

**Comment by ph** [comment]
Issues on CI doesn't seems related to the changes in this PR.

```
ok  	github.com/elastic/beats/x-pack/agent/pkg/sorted	1.067s	coverage: 100.0% of statements

=== RUN   TestTokenBucket

--- FAIL: TestTokenBucket (0.13s)

    token_bucket_test.go:39: 

        	Error Trace:	token_bucket_test.go:39

        	Error:      	Not equal: 

        	            	expected: 5

        	            	actual  : 4

        	Test:       	TestTokenBucket
```

**Comment by michalpristas** [comment]
no it's not related this is timing issue, this test fails on `TEST_ENVIRONMENT=0` quite often i did some refactoring of test but timing on flagged env is quite difficult

**Comment by ph** [comment]
> no it's not related this is timing issue, this test fails on TEST_ENVIRONMENT=0 quite often i did some refactoring of test but timing on flagged env is quite difficult

Let's create an issue to prioritize it.

File: x-pack/agent/pkg/agent/operation/monitoring.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+	// get output
+	config, err := getConfigFromStep(step)
+	if err != nil {
+		o.logger.Error("operator.getMonitoringSteps: getting config from step failed: %v", err)
+		return nil
+	}
+
+	outputIface, found := config[outputKey]
+	if !found {
+		o.logger.Errorf("operator.getMonitoringSteps: monitoring configuration not found for sidecar: %v", config)
```
[the remark this answers is not in the document]
Comment (michalpristas) [reply x-pack/agent/pkg/agent/operation/monitoring.go]: very good catch

File: x-pack/agent/pkg/core/plugin/app/monitoring/beats/beats_monitor.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+// Cleanup removes
+func (b *Monitor) Cleanup() error {
+	// do not cleanup logs, they might not be all processed
+	drop := b.monitoringDrop()
+	if drop == "" {
+		return nil
+	}
+
+	return os.RemoveAll(drop)
+}
```
[the remark this answers is not in the document]
Comment (michalpristas) [reply x-pack/agent/pkg/core/plugin/app/monitoring/beats/beats_monitor.go]: yes i will create an issue

File: x-pack/agent/pkg/core/plugin/app/monitoring/monitor.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+// NewMonitor creates a monitor based on a process configuration.
+func NewMonitor(isMonitorable bool, process, pipelineID string, downloadConfig *artifact.Config, monitorLogs, monitorMetrics bool) Monitor {
+	if !isMonitorable {
+		return noop.NewMonitor()
+	}
+
+	// so far we support only beats monitoring
+	return beats.NewMonitor(process, pipelineID, downloadConfig, monitorLogs, monitorMetrics)
+}
```
[the remark this answers is not in the document]
Comment (michalpristas) [reply x-pack/agent/pkg/core/plugin/app/monitoring/monitor.go]: with agent we will either create a new monitor and modify this factory or they will support beats interface and we do nothing. but this allow us to do both

File: x-pack/agent/pkg/core/plugin/app/monitoring/beats/beats_monitor.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	monitoringEndpoint string
+	loggingPath        string
+
+	monitorLogs    bool
+	monitorMetrics bool
+}
+
+// NewMonitor creates a beats monitor.
+func NewMonitor(process, pipelineID string, downloadConfig *artifact.Config, monitorLogs, monitorMetrics bool) *Monitor {
+	var monitoringEndpoint, loggingPath string
```
[the remark this answers is not in the document]
Comment (michalpristas) [reply x-pack/agent/pkg/core/plugin/app/monitoring/beats/beats_monitor.go]: path is common for multiple processes, in same directory there are multiple files e.g metricbeat, metricbeat.1... filebeat...name based on process

File: x-pack/agent/pkg/core/plugin/app/monitoring/beats/beats_monitor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (michalpristas) [reply x-pack/agent/pkg/core/plugin/app/monitoring/beats/beats_monitor.go]: #14820

**Comment by ph** [comment]
OK, code LGTM, I need to actually test/try this.

One thing we didn't consider is how we will monitor the Agent itself? Should we run the processes in a special streams?

**Comment by michalpristas** [comment]
if agent is connected to fleet we have a checkin api which contains subset of logs/events.

log-wise we dont monitor so far. simplest way would be to create a sidecar for an agent or agent sending it directly. which i guess is related to your latest issue

**Comment by ph** [comment]
Yes, we don't, lets create an issue to track it but having a sidecar to monitor the agents seems what we should do. in the end it's just matter of deciding in which streams the sidecards are created and inject the appropriate configuration.

**Comment by ph** [comment]
@michalpristas code LGTM, can we rebase? I've created a conflict because of multiples outputs  sorry about that. :(

**Comment by michalpristas** [comment]
this solution creates a monitoring for each output, so streams using output A are monitored by one set of sidecars which then sends logs to output A,
and streams using B output are monitored by different set sending results to B.

another option would be to create 
- DEFAULT output which will always be there and will be used also for monitoring
- `monitoring.output` option which will use monitoring 
^^ these will also mean that we will monitor all streams using single set of monitoring sidecars

cc @ph

**Comment by ph** [comment]
@michalpristas let's merge this as is and do a followup PR, but we will need to actually be able to specify the outputs using `monitoring.elasticsearch`, From what I've seen in the UI we can only specify a single monitoring per policy. @mattapperson or @nchaulet  can confirm that?

**Comment by nchaulet** [comment]
@ph yes you can just configure one monitoring per policy

**Comment by michalpristas** [comment]
i made the change to support `monitoring.elasticsearch`
look like it does not make sense to have it in a way i had it yesterday it is more complicated solution which does not comply to out contract.

**Pull request merged**